lib/auth/auth_with_roles.go

```diff
@@ -778,6 +781,81 @@ func (a *AuthWithRoles) DeleteWebSession(user string, sid string) error {
 	return a.authServer.DeleteWebSession(user, sid)
 }
 
+func (a *AuthWithRoles) GetAccessRequests(filter services.AccessRequestFilter) ([]services.AccessRequest, error) {
+	// An exception is made to allow users to get their own access requests.
+	if filter.User == "" || a.currentUserAction(filter.User) != nil {
+		if err := a.action(defaults.Namespace, services.KindAccessRequest, services.VerbList); err != nil {
+			return nil, trace.Wrap(err)
+		}
+		if err := a.action(defaults.Namespace, services.KindAccessRequest, services.VerbRead); err != nil {
+			return nil, trace.Wrap(err)
+		}
+	}
+	return a.authServer.GetAccessRequests(filter)
+}
+
+func (a *AuthWithRoles) WatchAccessRequests(ctx context.Context, filter services.AccessRequestFilter) (AccessRequestWatcher, error) {
+	// An exception is made to allow users to watch their own access requests.
+	if filter.User == "" || a.currentUserAction(filter.User) != nil {
+		if err := a.action(defaults.Namespace, services.KindAccessRequest, services.VerbList); err != nil {
+			return nil, trace.Wrap(err)
+		}
+		if err := a.action(defaults.Namespace, services.KindAccessRequest, services.VerbRead); err != nil {
+			return nil, trace.Wrap(err)
+		}
+	}
+	watcher, err := newAccessRequestWatcherFromEvents(ctx, a.authServer, filter)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return watcher, nil
+}
+
+func (a *AuthWithRoles) CreateAccessRequest(req services.AccessRequest) error {
+	// An exception is made to allow users to create access *pending* requests for themselves.
+	if !req.GetState().IsPending() || a.currentUserAction(req.GetUser()) != nil {
+		if err := a.action(defaults.Namespace, services.KindAccessRequest, services.VerbCreate); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+	// Ensure that an access request cannot outlive the identity that creates it.
+	if req.GetAccessExpiry().Before(a.authServer.GetClock().Now()) || req.GetAccessExpiry().After(a.identity.Expires) {
+		req.SetAccessExpiry(a.identity.Expires)
+	}
+	if err := a.authServer.CreateAccessRequest(req); err != nil {
+		return trace.Wrap(err)
+	}
+	err := a.authServer.EmitAuditEvent(events.AccessRequestCreated, events.EventFields{
+		events.AccessRequestID:    req.GetName(),
+		events.EventUser:          req.GetUser(),
+		events.UserRoles:          req.GetRoles(),
+		events.AccessRequestState: req.GetState().String(),
+	})
+	return trace.Wrap(err)
```

Comment:
This logic should be moved to the `AuthServer` struct. The only logic `AuthWithRoles` should have is:

1. Can the user perform this action?
2. Call something in `AuthServer` to perform action.

This way application logic and access logic are cleanly separated.